lib/SPIRV/SPIRVToOCL.cpp

```diff
@@ -0,0 +1,466 @@
+// Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimers in the documentation
+// and/or other materials provided with the distribution.
+// Neither the names of Advanced Micro Devices, Inc., nor the names of its
+// contributors may be used to endorse or promote products derived from this
+// Software without specific prior written permission.
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
+// THE SOFTWARE.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements common transform of SPIR-V builtins to OCL builtins.
+//
+// Some of the visit functions are translations to OCL2.0 builtins, but they
+// are currently used also for OCL1.2, so theirs implementations are placed
+// in this pass as a common functionality for both versions.
+//
+//===----------------------------------------------------------------------===//
+#define DEBUG_TYPE "spvtocl"
+
+#include "SPIRVToOCL.h"
+
+namespace SPIRV{
+
+static cl::opt<std::string>
+MangledAtomicTypeNamePrefix("spirv-atomic-prefix",
+  cl::desc("Mangled atomic type name prefix"),
+  cl::init("U7_Atomic"));
+
+static cl::opt<unsigned int>
+OCLBuiltinsVersion("spirv-ocl-builtins-version",
+  cl::desc("Specify version of OCL builtins to translate to (ex. 12, 20, 21, 22)"));
+
+char SPIRVToOCL::ID = 0;
+
+void SPIRVToOCL::visitCallInst(CallInst &CI) {
+  LLVM_DEBUG(dbgs() << "[visistCallInst] " << CI << '\n');
+  auto F = CI.getCalledFunction();
+  if (!F)
+    return;
+
+  auto MangledName = F->getName();
+  std::string DemangledName;
+  Op OC = OpNop;
+  if (!oclIsBuiltin(MangledName, &DemangledName) ||
+    (OC = getSPIRVFuncOC(DemangledName)) == OpNop)
+    return;
+  LLVM_DEBUG(dbgs() << "DemangledName = " << DemangledName.c_str() << '\n'
+    << "OpCode = " << OC << '\n');
+
+  if (OC == OpImageQuerySize || OC == OpImageQuerySizeLod) {
+    visitCallSPRIVImageQuerySize(&CI);
+    return;
+  }
+  if (OC == OpMemoryBarrier) {
+    visitCallSPIRVMemoryBarrier(&CI);
+    return;
+  }
+  if (isAtomicOpCode(OC)) {
+    visitCallSPIRVAtomicBuiltin(&CI, OC);
+    return;
+  }
+  if (isGroupOpCode(OC)) {
+    visitCallSPIRVGroupBuiltin(&CI, OC);
+    return;
+  }
+  if (isPipeOpCode(OC)) {
+    visitCallSPIRVPipeBuiltin(&CI, OC);
+    return;
+  }
+  if (OCLSPIRVBuiltinMap::rfind(OC))
+    visitCallSPIRVBuiltin(&CI, OC);
+}
+
+void SPIRVToOCL::visitCastInst(CastInst &Cast) {
+  if (!isa<ZExtInst>(Cast) && !isa<SExtInst>(Cast) && !isa<TruncInst>(Cast) &&
```

Comment:
Looks like it is easier to enumerate allowed casts then disallowed ones:

```c++
  if (isa<AddrSpacecastInst>(Cast) || isa<BitCastInst>(Cast) || isa<PtrToIntCast>(Cast) || isa<IntToPtrCast>(Cast))
    return;
```

Reply:
All of the code in this file is just copied from SPIRVToOCL20.cpp file, so it's not something that I wrote. Should we fix all of that things in this pull request?

Reply:
I don't have strong opinion on that. If you have time to improve things, it would be nice. Otherwise, you can skip my comments if code wasn't written by you right now.

But please apply clang-format: it is now a part of Travis CI checks

Reply:
Only if you want to I'd say... :-)  But please do that in a separate commit.

Reply:
Also applying the reformatting would be better in a separate commit as part of this pull request.

Reply:
Reformatting shouldn't be in a seperate comment I think. It's incorrect because of my IDE formatting settings different than clang-format, that not because original SPIRVToOCL20.cpp file was formatted incorrectly.

Reply:
Oh right, if the original formatting was fine then you should fix the formatting in the same commit indeed.